diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3cc06d57b..eca590b83 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,6 +35,7 @@ option(protobuf_BUILD_EXAMPLES "Build examples" OFF)
 option(protobuf_BUILD_PROTOBUF_BINARIES "Build protobuf libraries and protoc compiler" ON)
 option(protobuf_BUILD_PROTOC_BINARIES "Build libprotoc and protoc compiler" ON)
 option(protobuf_BUILD_LIBPROTOC "Build libprotoc" OFF)
+option(protobuf_BUILD_PYEXT "Build pyext" ON)
 option(protobuf_BUILD_LIBUPB "Build libupb" ON)
 option(protobuf_DISABLE_RTTI "Remove runtime type information in the binaries" OFF)
 option(protobuf_TEST_XML_OUTDIR "Output directory for XML logs from tests." "")
@@ -338,6 +339,10 @@ if (protobuf_BUILD_CONFORMANCE)
   include(${protobuf_SOURCE_DIR}/cmake/conformance.cmake)
 endif (protobuf_BUILD_CONFORMANCE)
 
+if(protobuf_BUILD_PYEXT)
+  include(${protobuf_SOURCE_DIR}/cmake/pyext.cmake)
+endif(protobuf_BUILD_PYEXT)
+
 if (protobuf_INSTALL)
   include(${protobuf_SOURCE_DIR}/cmake/install.cmake)
 endif (protobuf_INSTALL)
diff --git a/cmake/install.cmake b/cmake/install.cmake
index 0a31eade0..b08ba2c40 100644
--- a/cmake/install.cmake
+++ b/cmake/install.cmake
@@ -41,6 +41,10 @@ if (protobuf_BUILD_LIBUPB)
   list(APPEND _protobuf_libraries libupb)
 endif ()
 
+if(protobuf_BUILD_PYEXT)
+  list(APPEND _protobuf_libraries pyext)
+endif(protobuf_BUILD_PYEXT)
+
 foreach(_library ${_protobuf_libraries})
   if (UNIX AND NOT APPLE)
     set_property(TARGET ${_library}
@@ -52,7 +56,8 @@ foreach(_library ${_protobuf_libraries})
   install(TARGETS ${_library} EXPORT protobuf-targets
     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT ${_library}
     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT ${_library}
-    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT ${_library})
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT ${_library}
+    PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR} COMPONENT ${_library})
 endforeach()
 
 if (protobuf_BUILD_PROTOC_BINARIES)
@@ -190,3 +195,12 @@ if (protobuf_INSTALL_TESTS)
     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
 endif()
+
+# Install pyext headers
+if(protobuf_BUILD_PYEXT)
+  foreach(pyext_header ${pyext_headers})
+    file(RELATIVE_PATH relative_subdir ${pyext_dir} ${pyext_header})
+    get_filename_component(subdir ${relative_subdir} DIRECTORY)
+    install(FILES ${pyext_header} DESTINATION "include/google/protobuf/pyext/${subdir}")
+  endforeach()
+endif(protobuf_BUILD_PYEXT)
diff --git a/cmake/pyext.cmake b/cmake/pyext.cmake
new file mode 100644
index 000000000..65b1f1af1
--- /dev/null
+++ b/cmake/pyext.cmake
@@ -0,0 +1,31 @@
+cmake_policy(SET CMP0057 NEW)
+cmake_policy(SET CMP0074 NEW)
+
+find_package(Python3 COMPONENTS Development REQUIRED)
+
+set(pyext_dir ${CMAKE_SOURCE_DIR}/python/google/protobuf/pyext)
+message("pyext_dir ${pyext_dir}")
+file(GLOB_RECURSE pyext_files
+  LIST_DIRECTORIES false
+  ${pyext_dir}/*.cc
+  ${pyext_dir}/*.h
+)
+
+file(GLOB_RECURSE pyext_headers
+  LIST_DIRECTORIES false
+  ${pyext_dir}/*.h
+)
+
+if(MSVC)
+    # Ignore const string conversion and integer precision loss
+    add_compile_options(/Zc:strictStrings- /wd4244 /wd4267)
+endif()
+
+add_library(pyext STATIC ${pyext_files})
+target_include_directories(pyext PUBLIC
+    $<BUILD_INTERFACE:${protobuf_SOURCE_DIR}/python>
+    $<INSTALL_INTERFACE:include>
+#  ${pyext_dir}
+#  ${protobuf_source_dir}/python
+)
+target_link_libraries(pyext PUBLIC libprotobuf Python3::Python)
